# 单细胞课程-数据分析2

## 运行环境

```
Author:liuzihao
Date:2021-7-13
R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)
```

## 表达矩阵的质控（UMI）

当基因转化为表达矩阵后，需要对其进一步的指控，包括对表达量过低的基因删除、检测到基因过少的cell剔除等（上游分析中mapping和reads的QC未发现的细胞），减少分析过程中的noise

## 读取数据以及加载相关包

```r
library(SingleCellExperiment)
library(scater)
library(cowplot)
library(tidyverse)
options(stringsAsFactors = FALSE)
rm(list = ls())
setwd("D:\\Desktop\\单细胞学习\\单细胞课程数据")
molecules <- read.delim("molecules.txt",header = T)
ann <- read.delim("annotation.txt",header = T,encoding = "utf-8")
#对数据进行查看
> molecules[1:5,1:5]
                NA19098.r1.A01 NA19098.r1.A02 NA19098.r1.A03 NA19098.r1.A04 NA19098.r1.A05
ENSG00000237683              0              0              0              1              0
ENSG00000187634              0              0              0              0              0
ENSG00000188976              3              6              1              3              4
......
> head(ann)
  individual replicate well      batch      sample_id
1    NA19098        r1  A01 NA19098.r1 NA19098.r1.A01
2    NA19098        r1  A02 NA19098.r1 NA19098.r1.A02
3    NA19098        r1  A03 NA19098.r1 NA19098.r1.A03
......
```

## 数据质控

### 根据基因表达情况质控

- 首先创建`SCE`对象，首先删除没有在任何细胞中表达的基因，两种方法，`apply`和SCE提供的办法

```r
#创建SCE对象
umi <- SingleCellExperiment(assays = list(counts = as.matrix(molecules)),colData = ann)
#qc前的数据为19027个基因，864个cell
> dim(umi)
[1] 19027   864
##质控，首先使用apply的方法
molecules_keep <- molecules[apply(molecules,1,function(x){sum(x > 0) > 0}),]
#可以看出质控过后还有18726个基因保留
> dim(molecules_keep)
[1] 18726   864
##使用SCE自带的方法
#首先计算没有在任何细胞中表达的基因情况，返回一组逻辑值
keep_features <- rowSums(counts(umi) > 0) >0
#质质控
umi <- umi[keep_features,]
##可以看出两种方法质控的结果一致
> dim(umi)
[1] 18726   864
```

### 计算ERCC和线粒体在每个cell中的含量

- 教程年代久远，`SCE`中的`isSpike`和`calculateQCMetrics`均已经弃用，因此采取新的方法，首先使用`splitAltExps`对基因进行分组，具体用法如下，然后通过`perCellQCMetrics`计算每个指定的`feature`在细胞中的含量，但是`splitAltExps`仅仅可以计算一种指标

```r
##首先计算ERCC含量
#使用grepl函数返回每个基因是否为ERCC，返回值为一组与rownames()长度相同的逻辑值
is.spike <- grepl("^ERCC", rownames(umi))
#18726与rownames()长度一致
> table(is.spike)
is.spike
FALSE  TRUE 
18637    89 
#使用splitAltExps对基因进行分类(只能加一种进去？)
umi <- splitAltExps(umi, ifelse(is.spike, "ERCC_gene", "gene"))
#计算ERCC含量
perCellQCMetrics(umi)
> c
DataFrame with 864 rows and 6 columns
                     sum  detected altexps_ERCC_gene_sum altexps_ERCC_gene_detected altexps_ERCC_gene_percent     total
               <numeric> <numeric>             <numeric>                  <numeric>                 <numeric> <numeric>
NA19098.r1.A01     62135      8337                  1187                         31                   1.87455     63322
NA19098.r1.A02     62699      8203                  1277                         31                   1.99606     63976
NA19098.r1.A03     42509      7261                  1121                         28                   2.56933     43630
NA19098.r1.A04     52682      7955                  1240                         30                   2.29962     53922
NA19098.r1.A05     69625      8586                  1262                         33                   1.78030     70887
```

- 使用`perCellQCMetrics`计算QC指标，计算每个cell中的ERCC和线粒体含量，

  `subsets`: 指定基因列表，列表可以是字符串，也可以是逻辑值（TURE即为指定基因），**一定要指定每个基因集的名称**

  `flatten`: 是否简化输出结果，如果`=T`即简化输出结果，subsets_MT_sum subsets_MT_detected subsets_MT_percent ....列将以冒号分隔

```r
##作者提供了线粒体基因列表
MT <- c("ENSG00000198899", "ENSG00000198727", "ENSG00000198888",
        "ENSG00000198886", "ENSG00000212907", "ENSG00000198786",
        "ENSG00000198695", "ENSG00000198712", "ENSG00000198804",
        "ENSG00000198763", "ENSG00000228253", "ENSG00000198938",
        "ENSG00000198840")
#####方法1，以逻辑值的方式指定基因集
#返回每个基因是否为ERCC
is.spike <- grepl("^ERCC", rownames(umi))
#返回每个基因是否为线粒体基因
is.mt <- rownames(umi) %in% MT
#计算QC指标
perCellQCMetrics(umi,subsets  = list(MT = is.mt,ERCC=is.spike))
##解释
#sum：细胞发现的基因表达量的和
#detected：表达量大于0的基因和
#subsets_MT_sum: 线粒体基因表达量总和
#subsets_MT_detected：表达量大于0的线粒体基因的和
#subsets_MT_percent：线粒体基因占比
```

![image-20210713165344024](https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713165344024.png)

```r
#####方法1，以基因集的方式指定基因集
#获取ERCC和mt的基因列表
#获取ERCC基因列表
spike <- rownames(umi)[grep("ERCC", rownames(umi))]
#线粒体基因列表如上
#计算QC指标
perCellQCMetrics(umi,subsets  = list(MT = MT,ERCC=spike))
```

![image-20210713170256638](https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713170256638.png)

```r
#将QC指标的结果写入SCE对象
#colData(umi) <- cbind(umi@colData,perCellQCMetrics(umi,subsets  = list(MT = MT,ERCC=spike)))
umi <- addPerCellQC(umi,subsets  = list(MT = MT,ERCC=spike))	
#查看数据，已经写入
umi@colData
```

![image-20210713224347246](https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713224347246.png)

### 可视化查看每个质控指标的计算结果

- 每个cell中的基因表达量总和的分布情况，读取数/分子数较少的cell很可能被破坏或未能捕获细胞，因此应该被remove

```r
#查看每个cell检测到的基因数据分布
hist(x = umi@colData$sum,breaks = 100) 
#我们划定一个25000的阈值进行质控
filter_by_sum <- (umi$sum > 25000)
#可以看出来将会有46个cell被过滤掉
> table(filter_by_sum)
filter_by_sum
FALSE  TRUE 
   46   818 
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713174519038.png" alt="image-20210713174519038" style="zoom:80%;" />

- 查看每个cell中发现的基因数量的分布情况，从图中可以看出来，大多数细胞可以有7000 - 10000个基因被检测到，这个是scRNA-seq的正常水平。然而，这取决于实验方案和测序深度。例如，基于dropset的方法通常能检测到每个细胞更少的基因。上图中最显著的特征是分布左侧的“重尾”。如果各细胞的检出率相等，则分布应近似正态分布。因此，我们去除了分布尾部的那些细胞(检测到的少于7000个)

```r
hist(x = umi@colData$detected,breaks = 100)
#划定一个7000的阈值
filter_by_detected <- umi$detected > 7000
#可以看出有个116个cell将会被过滤
filter_by_detected <- umi$detected > 7000
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713180130804.png" alt="image-20210713180130804" style="zoom:80%;" />

### 根据线粒体基因含量和ERCC进行质控

- 通常可以使用ERCC来检测细胞的质量，通常如果ERCC的表达量显著过高，则说明细胞的其实RNA含量很低，说明cell可能存在着死亡或者降解

首先绘制**线粒体**和**count**数量的关系图

```r
plotColData(
  umi,
  x = "sum",
  y = "subsets_MT_percent",
  colour = "batch")
#划定一个10%的阈值
filter_by_MT <- umi$subsets_MT_percent <= 10
#可以看出有个31个cell将会被过滤
> table(filter_by_MT)
filter_by_MT
FALSE  TRUE 
   31   833
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713225125715.png" alt="image-20210713225125715" style="zoom:65%;" />

- 绘制counts数量和ERCC的关系, 下图可以看出 NA19098.r2批次的cell具有很高的ERCC基因表达

```r
plotColData(
  umi,
  x = "sum",
  x = "subsets_ERCC_percent",
  colour = "batch")
#移除ERCC过高的 NA19098.r2批次的cell
filter_by_ERCC <- umi$batch !=  "NA19098.r2"
table(filter_by_ERCC)
#将有96个cell被过滤掉
> table(filter_by_ERCC)
filter_by_ERCC
FALSE  TRUE 
   96   768 
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210713225727156.png" alt="image-20210713225727156" style="zoom:40%;" />

```
What would you expect to see in the ERCC vs counts plot if you were examining a dataset containing cells of different sizes (eg. normal & senescent cells)? 后者将会有更高的ERCC的表达
```

- 合并质控信息

```r
umi$use <- c(filter_by_sum & filter_by_detected & filter_by_ERCC & filter_by_MT)
> table(umi$use)
#可以看出将有207个cell被过滤掉
FALSE  TRUE 
  207   657
```

### PCA分析查看质控情况

- 使用`runPCA`和`plotReducedDim`查看数据

```r
#执行PCA分析，exprs_values为指定数据集，subset_row为指定使用哪些基因，默认采用前500个高变基因，ncomponents为保留的主成分数量
umi <- runPCA(umi,exprs_values = "counts",subset_row=rownames(umi)[1:nrow(umi)],ncomponents=25)
#绘制PCA图，dimred指定名称，具体查看reducedDimNames(umi)，查看存在哪些降维结果
plotReducedDim(umi,dimred = "PCA",  shape_by = "use",size_by = "sum",colour_by ="batch")
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714104741759.png" alt="image-20210714104741759" style="zoom:80%;" />

## ex：数据质控和数据可视化（reads）

### 读入reads数据并载入包

```r
library(SingleCellExperiment)
library(scater)
library(cowplot)
library(tidyverse)
setwd("D:\\Desktop\\单细胞学习\\单细胞课程数据")
rm(list = ls())
read <- read.delim("reads.txt",header = T)
ann <- read.delim("annotation.txt",header = T)
reads <- SingleCellExperiment(assays = list(counts = as.matrix(read)),colData = ann)
```

### 对基因进行质控---删除在所有cell中均没有表达的基因

```r
#计算在所有cell中均没有表达的基因
keep_gene_gr_0 <- apply(read,1,function(gene_count){sum(gene_count > 0) > 0})
reads <- reads[keep_gene_gr_0,]
```

### 计算质控指标---ERCC & MT

```r
#ERCC基因列表
ERCC <- rownames(reads)[grep("^ERCC-",rownames(reads))]
#线粒体基因列表
MT <-  c("ENSG00000198899", "ENSG00000198727", "ENSG00000198888",
         "ENSG00000198886", "ENSG00000212907", "ENSG00000198786",
         "ENSG00000198695", "ENSG00000198712", "ENSG00000198804",
         "ENSG00000198763", "ENSG00000228253", "ENSG00000198938",
         "ENSG00000198840")
#计算并将质控指标加入SCE对象
reads <- addPerCellQC(reads,subsets = list(MT = MT,ERCC = ERCC))
```

### 设定质控标准（cell）

```r
##质控标准的划定
hist(reads$detected,breaks = 100)
hist(reads$sum,breaks = 100)
p1 <- plotColData(reads,x="sum",y = "subsets_ERCC_percent",colour = "batch")
p2 <- plotColData(reads,x="sum",y = "subsets_MT_percent",colour = "batch")
plot_grid(p1, p2, labels = c("A", "B"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714112954057.png" alt="image-20210714112954057" style="zoom:70%;" />

<img src="D:\Desktop\s_note\data\picture\image-20210714113015571.png" alt="image-20210714113015571" style="zoom:70%;" />

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714115430246.png" alt="image-20210714115430246" style="zoom:80%;" />

```R
#质控标准划定
#每个cell的总reads数量大于1.3e6
#每个cell检测到的基因数大于7000
#删除离群的NA19098.r2批次，ERCC含量小于25
#线粒体含量小于30%
filter_by_total_counts <- reads$sum > 1.3e6
filter_by_total_detected <- reads$detected >7000
filter_by_ercc <- reads$batch != "NA19098.r2" & reads$subsets_ERCC_percent < 25
filter_by_mt <- reads$subsets_MT_percent < 30
#合并质控指标
fliter_qc <- filter_by_total_counts & filter_by_total_detected & filter_by_ercc & filter_by_mt
#258个cell将会被过滤
#table(filter_qc)
#FALSE  TRUE 
#  258   606 
```

### 设定质控标准（gene）

- 删除表达量过于低的基因

```r
#得到在符合标准的cell中表达量大于5的基因
keep_gene <- nexprs(reads[,reads$use],byrow = T,detection_limit = 1) > 5
#4039个基因被删除
> table(keep_gene)
keep_gene
FALSE  TRUE 
 4039 14687 
```

### 创建QC后的sce对象

```r
reads.qc <- reads[keep_gene,filter_qc]
#质控后还剩14687个基因和606个cell
>dim(reads.qc)
[1] 14687   606
```

```R
#基因表达的可视化
plotHighestExprs(reads, exprs_values = "counts")
```

### 数据的可视化

### 对质控前后的数据分别进行PCA分析

**原始数据**

```R
#首先将log转换后的数据（质控前后的数据）写入SCE对象
normcounts(reads) <- log2(counts(reads) + 1)
normcounts(reads.qc) <- log2(counts(reads.qc) + 1)
#分别对质控前后的数据进行PCA分析

#质控前数据PCA分析
reads <- runPCA(reads,exprs_values = "counts")
#颜色为是否需要丢弃的cell
pca_use <- plotReducedDim(reads,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
#颜色为批次
pca_batch <- plotReducedDim(reads,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

#质控后数据PCA分析
reads.qc <- runPCA(reads.qc,exprs_values = "counts",subset_row = rownames(reads.qc)[1:nrow(reads.qc)])
pca_qc_use <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
pca3_qc_batch <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#####可视化#####
plot_grid(pca_use,
          pca_batch,
          pca_qc_use,
          pca3_qc_batch,
          labels = c("pca_use","pca_batch","pca_qc_use","pca_qc_batch"))
```

![image-20210714134336269](https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714134336269.png)

**`log2`转换后的数据**

```r
#质控前数据PCA分析
reads <- runPCA(reads,exprs_values = "normcounts")
#颜色为是否需要丢弃的cell
pca_use <- plotReducedDim(reads,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
#颜色为批次
pca_batch <- plotReducedDim(reads,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

#质控后数据PCA分析
reads.qc <- runPCA(reads.qc,exprs_values = "normcounts",subset_row = rownames(reads.qc)[1:nrow(reads.qc)])
pca_qc_use <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
pca3_qc_batch <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#####可视化#####
plot_grid(pca_use,
          pca_batch,
          pca_qc_use,
          pca3_qc_batch,
          labels = c("pca_use","pca_batch","pca_qc_use","pca_qc_batch"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714135012512.png" alt="image-20210714135012512" style="zoom:80%;" />

### 对质控前后的数据分别进行t-sne分析

- 数据使用`log2`转换后的数据

关于perplexity [How to Use t-SNE Effectively (distill.pub)](https://distill.pub/2016/misread-tsne/)：  Perplexity parameter (should not be bigger than 3 * perplexity < nrow(X) - 1, see details for interpretation)

```r
#为了结果可重复，设定随机数种子
set.seed(123456)
reads <- runTSNE(reads,exprs_values = "normcounts",perplexity = 130,verbose	= T)
#颜色为是batch
plot_batch <- plotReducedDim(reads,dimred = "TSNE",colour_by = "batch",size_by = "sum",shape_by = "individual")
#颜色为是是否丢弃的cell
plot_use <- plotReducedDim(reads,dimred = "TSNE",colour_by = "use",size_by = "sum",shape_by = "individual")

reads.qc <- runTSNE(reads.qc,exprs_values = "normcounts",perplexity = 130)
plot.qc_batch <- plotReducedDim(reads.qc,dimred = "TSNE",colour_by = "batch",size_by = "sum",shape_by = "individual")
plot.qc_use <- plotReducedDim(reads.qc,dimred = "TSNE",colour_by = "use",size_by = "sum",shape_by = "individual")

plot_grid(plot_batch,plot_use,plot.qc_batch,plot.qc_use, labels = c("plot_batch", "plot_use","plot.qc_batch","plot.qc_use"))

```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714141149122.png" alt="image-20210714141149122" style="zoom:80%;" />

设定不同的`perplexity `, 查看对结果的影响，一般来说值越低会让cell间分的更开，更高的值cell间离得更近，默认值为细胞数除以5（四舍五入）

```
Furthermore tSNE requires you to provide a value of perplexity which reflects the number of neighbours used to build the nearest-neighbour network; a high value creates a dense network which clumps cells together while a low value makes the network more sparse allowing groups of cells to separate from each other. scater uses a default perplexity of the total number of cells divided by five (rounded down).
```

```r
list_PRE <- NULL
count = 0
for(pre in c(1,10,100,120,200)){
  count = count + 1
  print(str_c("t-sne ","perplexity=", pre))
  reads.qc <- runTSNE(reads.qc,exprs_values = "normcounts",perplexity = pre)
  list_PRE[[count]] <- plotReducedDim(reads.qc,dimred = "TSNE",colour_by = "batch",shape_by = "individual",size_by ="sum")
}


plot_grid(list_PRE[[1]],
          list_PRE[[2]],
          list_PRE[[3]],
          list_PRE[[4]],
          list_PRE[[5]],labels = c("perplexity=1","perplexity=10","perplexity=100","perplexity=120","perplexity=200"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714150450546.png" alt="image-20210714150450546" style="zoom:75%;" />



## ex：数据质控和数据可视化（UMI）

### 读入UMI数据并载入包

```r
library(SingleCellExperiment)
library(scater)
library(cowplot)
library(tidyverse)
setwd("D:\\Desktop\\单细胞学习\\单细胞课程数据")
rm(list = ls())
read <- read.delim("molecules.txt",header = T)
ann <- read.delim("annotation.txt",header = T)
```

### 对基因进行质控---删除在所有cell中均没有表达的基因

```r
#计算在所有cell中均没有表达的基因
keep_gene_gr_0 <- apply(read,1,function(gene_count){sum(gene_count > 0) > 0})
reads <- reads[keep_gene_gr_0,]
```

### 计算质控指标---ERCC & MT

```r
#ERCC基因列表
ERCC <- rownames(reads)[grep("^ERCC-",rownames(reads))]
#线粒体基因列表
MT <-  c("ENSG00000198899", "ENSG00000198727", "ENSG00000198888",
         "ENSG00000198886", "ENSG00000212907", "ENSG00000198786",
         "ENSG00000198695", "ENSG00000198712", "ENSG00000198804",
         "ENSG00000198763", "ENSG00000228253", "ENSG00000198938",
         "ENSG00000198840")
#计算并将质控指标加入SCE对象
reads <- addPerCellQC(reads,subsets = list(MT = MT,ERCC = ERCC))
```

### 设定质控标准（cell）

```r
##质控标准的划定
hist(reads$detected,breaks = 100)
hist(reads$sum,breaks = 100)
p1 <- plotColData(reads,x="sum",y = "subsets_ERCC_percent",colour = "batch")
p2 <- plotColData(reads,x="sum",y = "subsets_MT_percent",colour = "batch")
plot_grid(p1, p2, labels = c("A", "B"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714152735973.png" alt="image-20210714152735973" style="zoom:50%;" />

<img src="D:\Desktop\s_note\data\picture\image-20210714152757634.png" alt="image-20210714152757634" style="zoom:50%;" />

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714152815496.png" alt="image-20210714152815496" style="zoom:67%;" />

```R
#质控标准划定
#每个cell的总umi数量大于25000
#每个cell检测到的基因数大于7000
#删除离群的NA19098.r2批次，ERCC含量小于25
#线粒体含量小于10%
filter_by_total_counts <- reads$sum > 25000
filter_by_total_detected <- reads$detected >7000
filter_by_ercc <- reads$batch != "NA19098.r2" 
filter_by_mt <- reads$subsets_MT_percent < 10
#合并质控指标
fliter_qc <- filter_by_total_counts & filter_by_total_detected & filter_by_ercc & filter_by_mt
#207个cell将会被过滤
> table(filter_qc)
FALSE  TRUE 
  207   657 
```

### 设定质控标准（gene）

- 删除表达量过于低的基因

```r
#得到在符合标准的cell中表达量大于5的基因
keep_gene <- nexprs(reads[,reads$use],byrow = T,detection_limit = 1) > 5
#5964个基因会被删除
> table(keep_gene)
keep_gene
FALSE  TRUE 
 5964 12762 
```

### 创建QC后的sce对象

```r
reads$use <- fliter_qc
reads.qc <- reads[keep_gene,fliter_qc]
#质控后还剩14687个基因和606个cell
> dim(reads.qc)
[1] 12762   657
```

```R
#基因表达的可视化
plotHighestExprs(reads, exprs_values = "counts")
plotHighestExprs(reads.qc, exprs_values = "counts")
```

### 数据的可视化

### 对质控前后的数据分别进行PCA分析

**原始数据**

```R
#首先将log转换后的数据（质控前后的数据）写入SCE对象
normcounts(reads) <- log2(counts(reads) + 1)
normcounts(reads.qc) <- log2(counts(reads.qc) + 1)
#分别对质控前后的数据进行PCA分析

#质控前数据PCA分析
reads <- runPCA(reads,exprs_values = "counts")
#颜色为是否需要丢弃的cell
pca_use <- plotReducedDim(reads,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
#颜色为批次
pca_batch <- plotReducedDim(reads,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

#质控后数据PCA分析
reads.qc <- runPCA(reads.qc,exprs_values = "counts",subset_row = rownames(reads.qc)[1:nrow(reads.qc)])
pca_qc_use <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
pca3_qc_batch <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#####可视化#####
plot_grid(pca_use,
          pca_batch,
          pca_qc_use,
          pca3_qc_batch,
          labels = c("pca_use","pca_batch","pca_qc_use","pca_qc_batch"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714153428926.png" alt="image-20210714153428926" style="zoom:80%;" />

**`log2`转换后的数据**

```r
#质控前数据PCA分析
reads <- runPCA(reads,exprs_values = "normcounts")
#颜色为是否需要丢弃的cell
pca_use <- plotReducedDim(reads,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
#颜色为批次
pca_batch <- plotReducedDim(reads,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

#质控后数据PCA分析
reads.qc <- runPCA(reads.qc,exprs_values = "normcounts",subset_row = rownames(reads.qc)[1:nrow(reads.qc)])
pca_qc_use <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "use",size_by = "sum",shape_by = "individual")
pca3_qc_batch <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#####可视化#####
plot_grid(pca_use,
          pca_batch,
          pca_qc_use,
          pca3_qc_batch,
          labels = c("pca_use","pca_batch","pca_qc_use","pca_qc_batch"))
```

![image-20210714153118479](https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714153118479.png)

### 对质控前后的数据分别进行t-sne分析

- 数据使用`log2`转换后的数据

关于perplexity [How to Use t-SNE Effectively (distill.pub)](https://distill.pub/2016/misread-tsne/)：  Perplexity parameter (should not be bigger than 3 * perplexity < nrow(X) - 1, see details for interpretation)

```r
#为了结果可重复，设定随机数种子
set.seed(123456)
reads <- runTSNE(reads,exprs_values = "normcounts",perplexity = 130,verbose	= T)
#颜色为是batch
plot_batch <- plotReducedDim(reads,dimred = "TSNE",colour_by = "batch",size_by = "sum",shape_by = "individual")
#颜色为是是否丢弃的cell
plot_use <- plotReducedDim(reads,dimred = "TSNE",colour_by = "use",size_by = "sum",shape_by = "individual")

reads.qc <- runTSNE(reads.qc,exprs_values = "normcounts",perplexity = 130)
plot.qc_batch <- plotReducedDim(reads.qc,dimred = "TSNE",colour_by = "batch",size_by = "sum",shape_by = "individual")
plot.qc_use <- plotReducedDim(reads.qc,dimred = "TSNE",colour_by = "use",size_by = "sum",shape_by = "individual")

plot_grid(plot_batch,plot_use,plot.qc_batch,plot.qc_use, labels = c("plot_batch", "plot_use","plot.qc_batch","plot.qc_use"))

```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714153811906.png" alt="image-20210714153811906" style="zoom:80%;" />

设定不同的`perplexity `, 查看对结果的影响，一般来说值越低会让cell间分的更开，更高的值cell间离得更近，默认值为细胞数除以5（四舍五入）

```
Furthermore tSNE requires you to provide a value of perplexity which reflects the number of neighbours used to build the nearest-neighbour network; a high value creates a dense network which clumps cells together while a low value makes the network more sparse allowing groups of cells to separate from each other. scater uses a default perplexity of the total number of cells divided by five (rounded down).
```

```r
list_PRE <- NULL
count = 0
for(pre in c(1,10,100,120,200)){
  count = count + 1
  print(str_c("t-sne ","perplexity=", pre))
  reads.qc <- runTSNE(reads.qc,exprs_values = "normcounts",perplexity = pre)
  list_PRE[[count]] <- plotReducedDim(reads.qc,dimred = "TSNE",colour_by = "batch",shape_by = "individual",size_by ="sum")
}


plot_grid(list_PRE[[1]],
          list_PRE[[2]],
          list_PRE[[3]],
          list_PRE[[4]],
          list_PRE[[5]],labels = c("perplexity=1","perplexity=10","perplexity=100","perplexity=120","perplexity=200"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714153954560.png" alt="image-20210714153954560" style="zoom:80%;" />

## 批次效应（UMI）

### 确定引起批次效应的因子

在单细胞测序中有着非常多的干扰因子，因此从真正意义史昂复现单细胞的结果是一个挑战，本节我们将继续利用`scater`包继续探索批次效应

```r
#继续使用前面的数据
#首先查看批次效应，使用标准化后的数据的pca结果
#可以看出有一定的批次效应
pca3_qc_batch
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714212113161.png" alt="image-20210714212113161" style="zoom:80%;" />

```r
#查看各个变量可以解释的方差的百分比密度图
#横轴为可以解释的方差百分
?plotExplanatoryVariables
plotExplanatoryVariables(reads.qc,variables =c("individual","batch","subsets_MT_percent","subsets_ERCC_percent","sum","detected"),theme_size = 20 )
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210714215447917.png" alt="image-20210714215447917" style="zoom:80%;" />

从上图可以看出`detected`和`sum`都能解释较多的方差，而其他的元素`subsets_MT_percent`解释的较少，但是`batch`（批次）却具有很高的解释量，说明其中存在这一定的批次效应，除了上述因素还会有一些其他的因素造成批次效应,如细胞凋亡和不同的细胞时期，使用`[scLVM](https://github.com/PMBio/scLVM) `可以发现并去除他们

```r
#查看各个变量在各个主成分上可以解释的变量
plotExplanatoryPCs(reads.qc,variables = c("detected")) %>% .$data %>% arrange(desc(Pct_Var_Explained))  %>% head()
#可以看出detected可以在PC1上解释大部分变量
  PC Expl_Var Pct_Var_Explained
1  1 detected        9698.61629
2  2 detected          58.18530
3  9 detected          25.59999
4  8 detected          17.07818
5  5 detected          16.61502
6 11 detected          13.18313
plotExplanatoryPCs(reads.qc,variables = c("sum")) %>% .$data %>% arrange(desc(Pct_Var_Explained))  %>% head()
#sum可以在PC1上解释大部分变量
PC Expl_Var Pct_Var_Explained
1  1      sum       9541.117016
2  9      sum         96.072118
3 14      sum         57.372487
4 11      sum         42.096201
5  2      sum          9.277892
6 25      sum          8.504755
plotExplanatoryPCs(reads.qc,variables = c("batch")) %>% .$data %>% arrange(desc(Pct_Var_Explained))  %>% head()
#batch在很多轴（除了pc1）上都能有较高的解释度，因此批次效应存在
  PC Expl_Var Pct_Var_Explained
1  2    batch          9478.681
2  4    batch          9369.857
3  5    batch          9353.550
4  3    batch          9245.114
5  6    batch          8911.854
6  7    batch          7654.796
plotExplanatoryPCs(reads.qc,variables = c("subsets_MT_percent")) %>% .$data %>% arrange(desc(Pct_Var_Explained))  %>% head()
#线粒体的解释度均不高
  PC           Expl_Var Pct_Var_Explained
1  2 subsets_MT_percent         1804.1551
2  9 subsets_MT_percent         1144.8891
3  3 subsets_MT_percent          746.2923
4  1 subsets_MT_percent          692.0344
5 11 subsets_MT_percent          536.3850
6  7 subsets_MT_percent          534.1721
plotExplanatoryPCs(reads.qc,variables = c("subsets_ERCC_percent")) %>% .$data %>% arrange(desc(Pct_Var_Explained))  %>% head()
#ERCC的解释度均不高
  PC             Expl_Var Pct_Var_Explained
1  1 subsets_ERCC_percent         5428.9508
2  2 subsets_ERCC_percent         1475.8333
3  3 subsets_ERCC_percent          641.0234
4  4 subsets_ERCC_percent          582.8649
5  6 subsets_ERCC_percent          425.3028
6  5 subsets_ERCC_percent          121.8570
```

## 基因表达数据的标准化

### 各个标准化的方法

#### CPM

某一基因的表达量除以原始的基因表达量的总和后乘以100000000后得到的值为CPM值，需要将`ERCC`基因删除，只保留内源基因

不足之处：可能会掩盖一些表达量较高的基因

```r
#一个计算脚本
cpm <- function(counts){
  #删除内参，ERCC
  gene_nu <- nrow(counts)
  #print(gene_nu)
  counts <- counts[str_detect(pattern = "^ERCC",string = rownames(counts),negate = T),]
  gene_nu2 <- nrow(counts)
  #print(gene_nu2)
  print(str_c("Removed ERCC: ",gene_nu - gene_nu2))
  #计算CPM值
  cpm.counts <- apply(counts,2,function(x){x / sum(x)} * 1000000)
  cpm.counts <- log2(cpm.counts + 1) %>% as.data.frame()
  return(cpm.counts)
}
```

#### RLE标准化

由于要计算几何平均值，所以基因表达数据中不能有0的存在，所以其并不适合单细胞测序的结果，在. `edgeR`和 `scater` 中**RLE** 为`relative log expression`

首先第一步需要，计算**每个基因在所有cell中的表达水平的几何平均值**，得到每个基因的**几何平均值**，第二步，计算每个细胞的`size factor`每个细胞的`size factor`是基因表达与基因几何平均值比率的中位数。

```r
calc_sf <-
function (expr_mat, spikes = NULL) 
{
    geomeans <- exp(rowMeans(log(expr_mat[-spikes, ])))
    SF <- function(cnts) {
        median((cnts/geomeans)[(is.finite(geomeans) & geomeans > 
            0)])
    }
    norm_factor <- apply(expr_mat[-spikes, ], 2, SF)
    return(t(t(expr_mat)/norm_factor))
}
```

#### UQ标准化

首先每一列要除以该列的上四分位数（75%），通常计算完成后需要除以中位数以缩放表达水平，本方法的不足之处在于，很多时候单细胞测序结果细胞基因的表达的上四分位非常小（可能等于0），因此可以使用更大的分位数代替，如99%分位数

```r
calc_uq <-
function (expr_mat, spikes = NULL) 
{
    UQ <- function(x) {
        quantile(x[x > 0], 0.75)
    }
    uq <- unlist(apply(expr_mat[-spikes, ], 2, UQ))
    norm_factor <- uq/median(uq)
    return(t(t(expr_mat)/norm_factor))
}
```

#### TTM标准化

```
Another method is called TMM is the weighted trimmed mean of M-values (to the reference) proposed by (Robinson and Oshlack 2010). The M-values in question are the gene-wise log2 fold changes between individual cells. One cell is used as the reference then the M-values for each other cell is calculated compared to this reference. These values are then trimmed by removing the top and bottom ~30%, and the average of the remaining values is calculated by weighting them to account for the effect of the log scale on variance. Each non-reference cell is multiplied by the calculated factor. Two potential issues with this method are insufficient non-zero genes left after trimming, and the assumption that most genes are not differentially expressed.
```

#### scran标准化

```
scran package implements a variant on CPM specialized for single-cell data (L. Lun, Bach, and Marioni 2016). Briefly this method deals with the problem of vary large numbers of zero values per cell by pooling cells together calculating a normalization factor (similar to CPM) for the sum of each pool. Since each cell is found in many different pools, cell-specific factors can be deconvoluted from the collection of pool-specific factors using linear algebra.
```

#### Downsampling标准化

`Downsampling`标准化标准化是指，使得每个cell具有大致相同的分子数。该方法好处是会引入引入零值，从而不会改变检测到的基因数量。主要缺点是，该过程不是确定性的，所以每次`Downsampling`都会产生结果表达矩阵略有不同。因此，通常会在多个`Downsampling`上运行分析

```r
Down_Sample_Matrix <-
function (expr_mat) 
{
    min_lib_size <- min(colSums(expr_mat))
    down_sample <- function(x) {
        prob <- min_lib_size/sum(x)
        return(unlist(lapply(x, function(y) {
            rbinom(1, y, prob)
        })))
    }
    down_sampled_mat <- apply(expr_mat, 2, down_sample)
    return(down_sampled_mat)
}
```

#### 标准化方法的选择

**RLE, TMM, 以及UQ不适合单细胞数据**

### 分析比较不同标准化方法的结果

通过`PCA`方法和计算样品范围的 *relative log expression* (`scater::plotRLE()`)来比较不同标准化方法的效果。含有更多reads的细胞，其大部分基因的表达比所有细胞的中值表达水平也更高，得到RLE值为正值；含有更少reads的细胞，其大部分基因的表达比所有细胞的中值表达水平更低，得到的RLE为负值。而标准化后的`RLE`值应该为0。

```R
#RLE的计算
calc_cell_RLE <-
function (expr_mat, spikes = NULL) 
{
    RLE_gene <- function(x) {
        if (median(unlist(x)) > 0) {
            log((x + 1)/(median(unlist(x)) + 1))/log(2)
        }
        else {
            rep(NA, times = length(x))
        }
    }
    if (!is.null(spikes)) {
        RLE_matrix <- t(apply(expr_mat[-spikes, ], 1, RLE_gene))
    }
    else {
        RLE_matrix <- t(apply(expr_mat, 1, RLE_gene))
    }
    cell_RLE <- apply(RLE_matrix, 2, median, na.rm = T)
    return(cell_RLE)
}
```

#### 首先比较CPM标准化效果

```R
#首先构建删除了ERCC的数据
reads.qc_er <- reads.qc[str_detect(string = rownames(reads.qc),pattern = "^ERCC",negate = T),]
```

- PCA分析

```r
#计算cpm标准化值
logcounts(reads.qc_er) <- log2(calculateCPM(reads.qc_er) + 1)
#pca分析（cpm值）
reads.qc_er <- runPCA(reads.qc_er,exprs_values = "logcounts")
cpm1 <- plotReducedDim(reads.qc_er,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#使用为标准化的数据PCA
reads.qc_er <- runPCA(reads.qc_er,exprs_values = "normcounts")
no.cpm <- plotReducedDim(reads.qc_er,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#可视化
plot_grid(cpm1,no.cpm,labels = c("cpm","no.cpm"))
####
#可以看出来cpm标准化后的数据分群会更加明显一些
```

![image-20210715175954904](https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210715175954904.png)

- RLE分析

![image-20210716100940961](D:\Desktop\s_note\data\picture\image-20210716100940961.png)

```r
#rle分析
rle.cpm <- plotRLE(reads.qc_er, exprs_values = "logcounts",colour_by = "batch")
rle.nocpm <- plotRLE(reads.qc_er, exprs_values = "normcounts",colour_by = "batch")
plot_grid(rle.cpm,cpm1,rle.nocpm,no.cpm,labels = c("cpm.rle","cpm.pca","no.cpm","no.cpm.rle"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210716100747436.png" alt="image-20210716100747436" style="zoom:80%;" />

#### 比较scran包提供的标准化方法

- 首先根据表达矩阵对细胞进行聚类

```R
#min.size,每个cluster中含有的最少cell数量
qclust <- quickCluster(reads.qc, min.size = 30)
```

- 计算缩放因子`?computeSumFactors()`，并计算标准化后的矩阵

```R
reads.qc <- computeSumFactors(reads.qc,clusters = qclust,sizes = 15)
reads.qc <- logNormCounts(reads.qc)
```

- PCA分析与RLE-plot查看标准化效果

```r
#pca 分析
reads.qc <- runPCA(reads.qc,exprs_values = "normcounts")
no.scran.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
reads.qc <- runPCA(reads.qc,exprs_values = "logcounts")
scran.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
#plot RLE
scran.rle <- plotRLE(reads.qc,exprs_values = "logcounts",colour_by = "batch",size_by = "sum",shape_by = "individual")
no.scan.rle <- plotRLE(reads.qc,exprs_values = "normcounts",colour_by = "batch",size_by = "sum",shape_by = "individual")
#可视化
plot_grid(scran.pca,no.scran.pca,scran.rle,no.scan.rle,labels = c("scran.pca","no.scran.pca","scran.rle","no.scran.rle"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210719105616471.png" alt="image-20210719105616471" style="zoom:80%;" />

#### Downsampling标准化数据

- 使用前面提供的function计算Downsampling标准化后的数据

```r
#计算标准化后的数据
logcounts(reads.qc) <- log2(Down_Sample_Matrix(reads.qc@assays@data$counts) + 1)

reads.qc <- runPCA(reads.qc,exprs_values = "logcounts")
Downsampling.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
reads.qc <- runPCA(reads.qc,exprs_values = "normcounts")
no.Downsampling.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

Downsampling.rle <- plotRLE(reads.qc,exprs_values = "logcounts",colour_by = "batch",size_by = "sum",shape_by = "individual")
no.Downsampling.rle <- plotRLE(reads.qc,exprs_values = "normcounts",colour_by = "batch",size_by = "sum",shape_by = "individual")


plot_grid(Downsampling.pca,no.Downsampling.pca,Downsampling.rle,no.scan.rle,labels = c("Downsampling.pca","no.Downsampling.pca","Downsampling.rle","no.Downsampling.rle"))
```



<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210719110646765.png" alt="image-20210719110646765" style="zoom:80%;" />



### 根据基因长度和测序深度标准化数据

- **RPKM** - Reads Per Kilobase Million (for single-end sequencing)
- **FPKM** - Fragments Per Kilobase Million (same as **RPKM** but for paired-end sequencing, makes sure that paired ends mapped to the same fragment are not counted twice)
- **TPM** - Transcripts Per Kilobase Million (same as **RPKM**, but the order of normalizations is reversed - length first and sequencing depth second)

#### TPM标准化

首先对基因进行注释，拿到基因长度等信息，并且删除没有注释到的基因

```r
BiocManager::install("biomaRt")
#得到基因的注释信息
reads.qc <- getBMFeatureAnnos(
  reads.qc,
  filters = "ensembl_gene_id", 
  attributes = c(
    "ensembl_gene_id",
    "hgnc_symbol",
    "chromosome_name",
    "start_position",
    "end_position"
  ), 
  biomart = "ENSEMBL_MART_ENSEMBL", 
  dataset = "hsapiens_gene_ensembl",
  host = "www.ensembl.org"
)
##删除没有注释到的基因
reads.qc <- reads.qc[!is.na(rowData(reads.qc)$ensembl_gene_id), ]
##查看注释结果
> rowData(reads.qc)
DataFrame with 13332 rows and 5 columns
                ensembl_gene_id hgnc_symbol chromosome_name start_position end_position
                    <character> <character>     <character>      <integer>    <integer>
ENSG00000188976 ENSG00000188976       NOC2L               1         944203       959309
ENSG00000187961 ENSG00000187961      KLHL17               1         960584       965719
ENSG00000187608 ENSG00000187608       ISG15               1        1001138      1014540
```

计算各个基因的长度, 查看基因表达量和长度的关系

```r
#基因长度的计算
eff_length <- (rowData(reads.qc)$end_position - rowData(reads.qc)$start_position)/1000 %>% abs()
#查看基因长度和表达量的关系
plot(eff_length, rowMeans(counts(reads.qc)))
##二者没有任何关系，所以FPKM和TPM不适合本数据
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210719113157479.png" alt="image-20210719113157479" style="zoom:80%;" />

计算tpm值并查看pca和ple-plot

```r
#计算tpm值
tpm(reads.qc) <- log2(calculateTPM(reads.qc,eff_length) + 1)
#pca rle-plot以及可视化
reads.qc <- runPCA(reads.qc,exprs_values = "tpm")
tpm.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

reads.qc <- runPCA(reads.qc,exprs_values = "normcounts")
no.tpm.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

tpm.rle <- plotRLE(reads.qc,exprs_values = "tpm",colour_by = "batch",size_by = "sum",shape_by = "individual")
no.tpm.rle <- plotRLE(reads.qc,exprs_values = "normcounts",colour_by = "batch",size_by = "sum",shape_by = "individual")

plot_grid(tpm.pca,no.tpm.pca,tpm.rle,no.scan.rle,labels = c("tpm.pca","no.tpm.pca","tpm.rle","no.tpm.rle"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210719113813977.png" alt="image-20210719113813977" style="zoom:80%;" />

#### FPKM标准化

计算FPKM值并查看pca和ple-plot

```r
fpkm(reads.qc) <- log2(calculateFPKM(reads.qc,eff_length) + 1)

reads.qc <- runPCA(reads.qc,exprs_values = "fpkm")
fpkm.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")
reads.qc <- runPCA(reads.qc,exprs_values = "normcounts")
no.fpkm.pca <- plotReducedDim(reads.qc,dimred = "PCA",colour_by = "batch",size_by = "sum",shape_by = "individual")

fpkm.rle <- plotRLE(reads.qc,exprs_values = "fpkm",colour_by = "batch",size_by = "sum",shape_by = "individual")
no.fpkm.rle <- plotRLE(reads.qc,exprs_values = "normcounts",colour_by = "batch",size_by = "sum",shape_by = "individual")

plot_grid(fpkm.pca,no.fpkm.pca,fpkm.rle,no.scan.rle,labels = c("fpkm.pca","no.fpkm.pca","fpkm.rle","no.fpkm.rle"))
```

<img src="https://aironi.oss-cn-beijing.aliyuncs.com/typro_image/image-20210719114341464.png" alt="image-20210719114341464" style="zoom:80%;" />

FPKM的结果与CPM相似，是由于二者的区别只要是否使用基因长度进行了矫正，前者计算了cpm值而后用基因长度矫正，TPM是先使用基因长度矫正，而后计算cpm所以与cpm的结果差异较大

```
Note The PCA looks for differences between cells. Gene length is the same across cells for each gene thus FPKM is almost identical to the CPM plot (it is just rotated) since it performs CPM first then normalizes gene length. Whereas, TPM is different because it weights genes by their length before performing CPM.
```

### 批次效应的处理

w

